File: C++/2 Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
using namespace std;

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *dummy = new ListNode(-1);
        auto cur = dummy;
        auto p = l1, q = l2;
        int t = 0;
        while (p || q || t > 0)
        {
            if (p)
            {
                t += p->val;
                p = p->next;
            }
            if (q)
            {
                t += q->val;
                q = q->next;
            }
            cur->next = new ListNode(t % 10);
            cur = cur->next;
            t /= 10;
        }
        return dummy->next;
    }
};

File: Swift/2 Add Two Numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var p = l1, q = l2
        var dummy: ListNode = .init(-1)
        var cur = dummy
        var t = 0
        while p != nil || q != nil || t > 0 {
            if let tmp = p {
                p = tmp.next
                t += tmp.val
            }
            if let tmp = q {
                q = tmp.next
                t += tmp.val
            }

            var tmp: ListNode? = .init(t % 10)
            cur.next = tmp
            cur = tmp!
            t /= 10
        }
        return dummy.next
    }
}
